SOSS/dms/soss_centroids.py

```diff
@@ -0,0 +1,377 @@
+        comb = slopevals - zero_roll(slopevals, -width)
+
+        # Find the maximum resulting slope.
+        args = np.nanargmax(comb, axis=0)
+        vals = np.nanmax(comb, axis=0)
+
+        # Update the best values.
+        mask = (vals > slopes_best)
+        slopes_best = np.where(mask, vals, slopes_best)
+        ytrace_best = np.where(mask, ypix[args], ytrace_best)
+        widths_best = np.where(mask, width, widths_best)
+
+    # Set the y position to NaN if the best slope was zero.
+    ytrace_best = np.where(slopes_best != 0, ytrace_best + widths_best/2., np.nan)
+    widths_best = np.where(slopes_best != 0, widths_best, np.nan)
+
+    if verbose:
+
+        plt.imshow(image, origin='lower', cmap='inferno', norm=colors.LogNorm())
+        plt.plot(ytrace_min, lw=2, ls='--', c='black')
+        plt.plot(ytrace_max, lw=2, ls='--', c='black')
+        plt.plot(ytrace_best, lw=2, c='black')
+
+        plt.show()
+        plt.close()
+
+    return ytrace_max, ytrace_min, ytrace_best, widths_best
+
+
+def get_uncontam_centroids_edgetrig(image, header=None, mask=None, poly_order=11,
+                                    halfwidth=5, mode='combined', verbose=False):
+    """Determine the x, y coordinates of the trace using the derivatives along the y-axis.
+    Works for either order if there is no contamination.
+
+    :param image: A 2D image of the detector.
+    :param header: The header from one of the SOSS reference files.
+    :param mask: A boolean array of the same shape as stack. Pixels corresponding to True values will be masked.
+    :param poly_order: Order of the polynomial to fit to the extracted trace positions.
+    :param halfwidth: the size of the window used when computing the derivatives.
+    :param mode: Which trace values to use. Can be 'maxedge', 'minedge', 'mean' or 'combined'.
+    :param verbose: If set True some diagnostic plots will be made.
+
+    :type image: array[float]
+    :type header: astropy.io.fits.Header
+    :type mask: array[bool]
+    :type poly_order: int
+    :type halfwidth: int
+    :type mode: str
+    :type verbose: bool
+
+    :returns: xtrace, ytrace, tracewidth, param - The x, y coordinates of trace as computed from the best fit polynomial
+    and the best-fit polynomial parameters.
+    :rtype: Tuple(array[float], array[float], array[float])
+    """
+
+    # If no mask was given use all pixels.
+    if mask is None:
+        mask = np.zeros_like(image, dtype='bool')
+
+    # Call the script that determines the dimensions of the stack.
+    dimx, dimy, xos, yos, xnative, ynative, padding, refpix_mask = \
+        get_image_dim(image, header=header, verbose=verbose)
+
+    # Replace masked pixel values with NaNs.
+    image_masked = np.where(mask | ~refpix_mask, np.nan, image)
+
+    # Use edge trigger to compute the edges and center of the trace.
+    ytrace_max, ytrace_min, ytrace_best, widths_best = edge_trigger(image_masked, halfwidth=halfwidth, yos=yos, verbose=verbose)
```

Comment:
Lovely long line. Again personal opinion but I prefer:
```python
yopts = edge_trigger(image_masked, halfwidth=halfwidth, yos=yos, verbose=verbose)
ytrace_max, ytrace_min, ytrace_best, widths_best  = yopts
```
or 
```python
ykwargs = dict(halfwidth=halfwidth,  yos=yos, verbose=verbose)
ytrace_max, ytrace_min, ytrace_best, widths_best  = edge_trigger(image_masked, **ykwargs)
```
or both if a very long line:
```python
ykwargs = dict(halfwidth=halfwidth,  yos=yos, verbose=verbose)
yopts = edge_trigger(image_masked, **ykwargs)
ytrace_max, ytrace_min, ytrace_best, widths_best  = yopts
```

Reply:
I prefer the kwargs method as a first resort I think.